src/snapshot/mod.rs

```diff
@@ -0,0 +1,232 @@
+use crate::{ConfirmedFrameCount, Rollback, DEFAULT_FPS};
+use bevy::{prelude::*, utils::HashMap};
+use std::{collections::VecDeque, marker::PhantomData};
+
+mod checksum;
+mod component_checksum_hash;
+mod component_clone;
+mod component_copy;
+mod component_map;
+mod component_reflect;
+mod entity;
+mod resource_checksum_hash;
+mod resource_clone;
+mod resource_copy;
+mod resource_map;
+mod resource_reflect;
+mod rollback_entity_map;
+mod set;
+
+pub use checksum::*;
+pub use component_checksum_hash::*;
+pub use component_clone::*;
+pub use component_copy::*;
+pub use component_map::*;
+pub use component_reflect::*;
+pub use entity::*;
+pub use resource_checksum_hash::*;
+pub use resource_clone::*;
+pub use resource_copy::*;
+pub use resource_map::*;
+pub use resource_reflect::*;
+pub use rollback_entity_map::*;
+pub use set::*;
+
+pub mod prelude {
+    pub use super::{
+        Checksum, GgrsChecksumPlugin, GgrsComponentChecksumHashPlugin,
+        GgrsComponentMapEntitiesPlugin, GgrsComponentSnapshotClonePlugin,
+        GgrsComponentSnapshotCopyPlugin, GgrsComponentSnapshotReflectPlugin,
+        GgrsEntitySnapshotPlugin, GgrsResourceChecksumHashPlugin, GgrsResourceMapEntitiesPlugin,
+        GgrsResourceSnapshotClonePlugin, GgrsResourceSnapshotCopyPlugin,
+        GgrsResourceSnapshotReflectPlugin, GgrsSnapshotSetPlugin, LoadWorldSet, SaveWorldSet,
+    };
+}
+
+/// Typical [`Resource`] used to store snapshots for a [`Resource`] `R` as the type `As`.
+/// For most types, the default `As = R` will suffice.
+pub type GgrsResourceSnapshots<R, As = R> = GgrsSnapshots<R, Option<As>>;
+
+/// Typical [`Resource`] used to store snapshots for a [`Component`] `C` as the type `As`.
+/// For most types, the default `As = C` will suffice.
+pub type GgrsComponentSnapshots<C, As = C> = GgrsSnapshots<C, GgrsComponentSnapshot<C, As>>;
+
+/// Collection of snapshots for a type `For`, stored as `As`
+#[derive(Resource)]
+pub struct GgrsSnapshots<For, As = For> {
+    /// Queue of snapshots, newest at the front, oldest at the back.
+    snapshots: VecDeque<As>,
+    /// Queue of snapshots, newest at the front, oldest at the back.\
+    /// Separate from snapshots to avoid padding.
+    frames: VecDeque<i32>,
+    /// Maximum amount of snapshots to store at any one time
+    depth: usize,
+    _phantom: PhantomData<For>,
+}
+
+impl<For, As> Default for GgrsSnapshots<For, As> {
+    fn default() -> Self {
+        Self {
+            snapshots: VecDeque::with_capacity(DEFAULT_FPS),
+            frames: VecDeque::with_capacity(DEFAULT_FPS),
+            depth: DEFAULT_FPS, // TODO: Make sensible choice here
```

Comment:
It definitely should be decided you're right. I think `DEFAULT_FPS` is a good default value, since I think it's important that snapshots not be collected infinitely if (somehow) a confirmed frame is never reached. But I'm open to other opinions on this one, as I have somewhat arbitrarily just decided that 1 second of rollback is a good maximum.

Reply:
It is impossible for GGRS to step ahead of the confirmed frame by more than `max_prediction_window` (parameter in the session, default is 8). You should use the parameter from the session to construct this.

Reply:
Of course that makes much more sense! I've set the default value to 8, and I've updated the `discard_old_snapshots` system to also check if the max prediction window has changed, and use that as their depth.